homeassistant/components/http/__init__.py

```diff
@@ -329,34 +338,103 @@ async def serve_file(request: web.Request) -> web.FileResponse:
+            if context is None:
+                raise HomeAssistantError(
+                    "Failed to create ssl context, no fallback available because a peer certificate is required."
                 )
 
-        else:
-            context = None
+            context.verify_mode = ssl.CERT_REQUIRED
+            context.load_verify_locations(self.ssl_peer_certificate)
+
+        return context, False
+
+    def _create_emergency_ssl_context(self) -> ssl.SSLContext | None:
+        """Create an emergency ssl certificate so we can still startup."""
+        context = ssl_util.server_context_modern()
+        host: str
+        try:
+            host = cast(str, URL(get_url(self.hass, prefer_external=True)).host)
+        except NoURLAvailableError:
+            host = "homeassistant.local"
+        key = rsa.generate_private_key(
+            public_exponent=65537,
+            key_size=2048,
+        )
+        subject = issuer = x509.Name(
+            [
+                x509.NameAttribute(
+                    NameOID.ORGANIZATION_NAME, "Home Assistant Emergency Certificate"
+                ),
+                x509.NameAttribute(NameOID.COMMON_NAME, host),
+            ]
+        )
+        cert = (
+            x509.CertificateBuilder()
+            .subject_name(subject)
+            .issuer_name(issuer)
+            .public_key(key.public_key())
+            .serial_number(x509.random_serial_number())
+            .not_valid_before(datetime.datetime.utcnow())
+            .not_valid_after(datetime.datetime.utcnow() + datetime.timedelta(days=30))
+            .add_extension(
+                x509.SubjectAlternativeName([x509.DNSName(host)]),
+                critical=False,
+            )
+            .sign(key, hashes.SHA256())
+        )
+        with NamedTemporaryFile() as cert_pem, NamedTemporaryFile() as key_pem:
+            cert_pem.write(cert.public_bytes(serialization.Encoding.PEM))
+            key_pem.write(
+                key.private_bytes(
+                    serialization.Encoding.PEM,
+                    format=serialization.PrivateFormat.TraditionalOpenSSL,
+                    encryption_algorithm=serialization.NoEncryption(),
+                )
+            )
+            cert_pem.flush()
+            key_pem.flush()
+            context.load_cert_chain(cert_pem.name, key_pem.name)
+        return context
+
+    async def start(self) -> None:
+        """Start the aiohttp server."""
+        context: ssl.SSLContext | None = None
+        if self.ssl_certificate:
+            context, is_emergency_cert = await self.hass.async_add_executor_job(
+                self._create_ssl_context
+            )
+            if is_emergency_cert:
+                _LOGGER.critical(
+                    "Home Assistant is running in safe mode with an emergency self signed ssl certificate because the configured SSL certificate was not usable"
+                )
+                self.hass.config.safe_mode = True
```

Comment:
I had imagined we would rely on triggering a safe mode via the logic in bootstrap: https://github.com/home-assistant/core/blob/dev/homeassistant/bootstrap.py#L148-L152

It is automatically activated if `frontend` fails to load, which as a dependency on `http`. So if the `http` integration fails to set up, safe mode is activated.

Reply:
So something like, force failure if ssl can't be created, but once safe_mode is passed in to the config, create the self signed and startup ?

Reply:
That gets a bit messy because of the delayed start of http... will take a shot at it

Reply:
Hmmm. the webserver is up and running and it started up in safe mode but I can't do anything to recover as this is all I get
<img width="887" alt="Screen Shot 2022-02-18 at 16 06 12" src="https://user-images.githubusercontent.com/663432/154767693-dd4596eb-bafc-4176-a6e0-8fa614643d89.png">

Reply:
So if `http` fails

```
2022-02-18 16:08:33 ERROR (MainThread) [homeassistant.setup] Unable to set up dependencies of api. Setup failed for dependencies: http
2022-02-18 16:08:33 ERROR (MainThread) [homeassistant.setup] Setup failed for api: (DependencyError(...), 'Could not setup dependencies: http')
2022-02-18 16:08:33 ERROR (MainThread) [homeassistant.setup] Unable to set up dependencies of diagnostics. Setup failed for dependencies: http
2022-02-18 16:08:33 ERROR (MainThread) [homeassistant.setup] Setup failed for diagnostics: (DependencyError(...), 'Could not setup dependencies: http')
2022-02-18 16:08:33 ERROR (MainThread) [homeassistant.setup] Unable to set up dependencies of websocket_api. Setup failed for dependencies: http
2022-02-18 16:08:33 ERROR (MainThread) [homeassistant.setup] Setup failed for websocket_api: (DependencyError(...), 'Could not setup dependencies: http')
2022-02-18 16:08:33 ERROR (MainThread) [homeassistant.setup] Unable to set up dependencies of webhook. Setup failed for dependencies: http
2022-02-18 16:08:33 ERROR (MainThread) [homeassistant.setup] Setup failed for webhook: (DependencyError(...), 'Could not setup dependencies: http')
2022-02-18 16:08:33 ERROR (MainThread) [homeassistant.setup] Unable to set up dependencies of config. Setup failed for dependencies: http
2022-02-18 16:08:33 ERROR (MainThread) [homeassistant.setup] Setup failed for config: (DependencyError(...), 'Could not setup dependencies: http')
2022-02-18 16:08:33 ERROR (MainThread) [homeassistant.setup] Unable to set up dependencies of image. Setup failed for dependencies: http
2022-02-18 16:08:33 ERROR (MainThread) [homeassistant.setup] Setup failed for image: (DependencyError(...), 'Could not setup dependencies: http')
2022-02-18 16:08:33 ERROR (MainThread) [homeassistant.setup] Unable to set up dependencies of auth. Setup failed for dependencies: http
2022-02-18 16:08:33 ERROR (MainThread) [homeassistant.setup] Setup failed for auth: (DependencyError(...), 'Could not setup dependencies: http')
2022-02-18 16:08:33 ERROR (MainThread) [homeassistant.setup] Unable to set up dependencies of search. Setup failed for dependencies: websocket_api
2022-02-18 16:08:33 ERROR (MainThread) [homeassistant.setup] Setup failed for search: (DependencyError(...), 'Could not setup dependencies: websocket_api')
2022-02-18 16:08:33 ERROR (MainThread) [homeassistant.setup] Unable to set up dependencies of analytics. Setup failed for dependencies: api, websocket_api
2022-02-18 16:08:33 ERROR (MainThread) [homeassistant.setup] Setup failed for analytics: (DependencyError(...), 'Could not setup dependencies: api, websocket_api')
2022-02-18 16:08:33 ERROR (MainThread) [homeassistant.setup] Unable to set up dependencies of network. Setup failed for dependencies: websocket_api
2022-02-18 16:08:33 ERROR (MainThread) [homeassistant.setup] Setup failed for network: (DependencyError(...), 'Could not setup dependencies: websocket_api')
2022-02-18 16:08:33 ERROR (MainThread) [homeassistant.setup] Unable to set up dependencies of usb. Setup failed for dependencies: websocket_api
2022-02-18 16:08:33 ERROR (MainThread) [homeassistant.setup] Setup failed for usb: (DependencyError(...), 'Could not setup dependencies: websocket_api')
2022-02-18 16:08:33 ERROR (MainThread) [homeassistant.setup] Unable to set up dependencies of cloud. Setup failed for dependencies: http, webhook
2022-02-18 16:08:33 ERROR (MainThread) [homeassistant.setup] Setup failed for cloud: (DependencyError(...), 'Could not setup dependencies: http, webhook')
2022-02-18 16:08:33 ERROR (MainThread) [homeassistant.setup] Unable to set up dependencies of person. Setup failed for dependencies: image
2022-02-18 16:08:33 ERROR (MainThread) [homeassistant.setup] Setup failed for person: (DependencyError(...), 'Could not setup dependencies: image')
2022-02-18 16:08:33 ERROR (MainThread) [homeassistant.setup] Unable to set up dependencies of ssdp. Setup failed for dependencies: network
2022-02-18 16:08:33 ERROR (MainThread) [homeassistant.setup] Setup failed for ssdp: (DependencyError(...), 'Could not setup dependencies: network')
2022-02-18 16:08:33 ERROR (MainThread) [homeassistant.setup] Unable to set up dependencies of zeroconf. Setup failed for dependencies: network, api
2022-02-18 16:08:33 ERROR (MainThread) [homeassistant.setup] Setup failed for zeroconf: (DependencyError(...), 'Could not setup dependencies: network, api')
2022-02-18 16:08:33 ERROR (MainThread) [homeassistant.setup] Unable to set up dependencies of onboarding. Setup failed for dependencies: analytics, auth, http, person
2022-02-18 16:08:33 ERROR (MainThread) [homeassistant.setup] Setup failed for onboarding: (DependencyError(...), 'Could not setup dependencies: analytics, auth, http, person')
2022-02-18 16:08:33 ERROR (MainThread) [homeassistant.setup] Unable to set up dependencies of frontend. Setup failed for dependencies: api, auth, config, diagnostics, http, onboarding, search, websocket_api
2022-02-18 16:08:33 ERROR (MainThread) [homeassistant.setup] Setup failed for frontend: (DependencyError(...), 'Could not setup dependencies: api, auth, config, diagnostics, http, onboarding, search, websocket_api')
```

All the deps fail and then we don't try again .

Reply:
Well I got it up and running but all the apis don't work since they aren't being loaded again

```

    if DATA_LOGGING in hass.data:
        hass.http.register_view(APIErrorLog)
```

Reply:
<img width="924" alt="Screen Shot 2022-02-18 at 16 41 49" src="https://user-images.githubusercontent.com/663432/154770944-4b666d08-e6e3-482b-ba80-2f8232b6e6ee.png">